types/src/generator/index.ts

```diff
@@ -144,39 +152,167 @@ function collectClasses(map: StructureMap): Set<string> {
+      if (globalIncluded.has(name) && !nestedTypeNames.has(name)) continue;
+
+      // If this structure was exported by another module, import it. Note we
+      // don't need to check whether we've already imported the type as
+      // `moduleIncluded` is a `Set`.
+      const maybeOwningModule = typeExports.get(name);
+      if (maybeOwningModule !== undefined && maybeOwningModule !== specifier) {
+        // Internal modules only have default exports, so we generate something
+        // that looks like this:
+        // ```
+        // import _internal1 from "node-internal:async_hooks";
+        // import AsyncLocalStorage = _internal1.AsyncLocalStorage; // (type & value alias)
+        // ```
+        const identifier = f.createIdentifier(`_internal${nextImportId++}`);
+        const importClause = f.createImportClause(
+          false,
+          /* name */ identifier,
+          /* namedBindings */ undefined
+        );
+        const importDeclaration = f.createImportDeclaration(
+          /* modifiers */ undefined,
+          importClause,
+          f.createStringLiteral(maybeOwningModule)
+        );
+        const typeName = getTypeName(name);
+        const importEqualsDeclaration = f.createImportEqualsDeclaration(
+          /* modifiers */ undefined,
+          /* isTypeOnly */ false,
+          typeName,
+          f.createQualifiedName(identifier, typeName)
+        );
+        statements.unshift(importDeclaration, importEqualsDeclaration);
+
+        continue;
+      }
+
+      // Otherwise, just include the structure in the module
+      const structure = structureMap.get(name);
+      assert(structure !== undefined);
+      const asClass = classes.has(name);
+      const statement = createStructureNode(structure, {
+        asClass,
+        ambientContext: true,
+        // nameOverride: nestedNameOverrides.get(name), // TODO: remove
+      });
+      statements.push(statement);
+    }
 
-  // Log ignored types to make sure we didn't forget anything
-  if (ignored.length > 0) {
-    console.warn(
-      "WARNING: The following types were not referenced from any `JSG_TS_ROOT()`ed type and have been omitted from the output. " +
-        "This could be because of disabled compatibility flags."
+    const moduleBody = f.createModuleBlock(statements);
+    const moduleDeclaration = f.createModuleDeclaration(
+      [f.createToken(ts.SyntaxKind.DeclareKeyword)],
+      f.createStringLiteral(specifier),
+      moduleBody
     );
-    for (const name of ignored) console.warn(`- ${name}`);
-  }
+    flatNodes.push(moduleDeclaration);
+  });
+
+  return { nodes: flatNodes, structureMap };
+}
+
+export function collectTypeScriptModules(root: StructureGroups): string {
+  let result = "";
+
+  root.getModules().forEach((module) => {
+    if (!module.isTsDeclarations()) return;
+    const declarations = module
+      .getTsDeclarations()
+      .replaceAll(/^\/\/\/.+$/gm, (match) => {
```

Comment:
nit: a comment explaining what this regex is doing would be helpful for future generations

Reply:
https://github.com/cloudflare/workerd/pull/2379/commits/409f10dca728a04a7800f20173a96aa75a4ae1b8